docs/lesson-plans/week-4.md

```diff
@@ -0,0 +1,612 @@
+
+##### Client Errors
+
+These represent errors that a client has made and are used to let it know what went wrong.
+
+Here are a few of the common ones:
+
+| Code | Name | Description |
+| :--- | :--- | :--- |
+| 400 | Bad Request | There was an issue with the request (usually due to invalid syntax) | 
+| 401 | Unauthorized | The client needs to authenticate (log in) before using this resource |
+| 403 | Forbidden | The client is authorized, but doesn't have the right permissions |
+| 404 | Not Found | The requested resource could not be found or the URL is not recognized |
+| 405 | Method Not Allowed | The wrong method is being used for a route (like a POST vs a GET, will be explained later) |
+| 410 | Gone | The requested resource has been permanently deleted from the server |
+| 418 | I'm a teapot | The server refuses the attempt to brew coffee (yes this is real) | 
+
+##### Server Errors
+
+These responses let the client know that there was an issue processing the request by the server. Usually you want to 
+avoid returning these because you properly handle every error that can pop up, but there are times when it's unavoidable 
+or necessary.
+
+Here are a few of the common ones:
+
+| Code | Name | Description |
+| :--- | :--- | :--- |
+| 500 | Internal Server Error | Usually thrown on an exception the server doesn't know how to handle, you want to try and avoid these |
+| 502 | Bad Gateway | The server (acting as a gateway) got an invalid response (usually this happens for certain server configuration errors) |
+| 503 | Service Unavailable | The server isn't ready to handle the request (usually when the server is down or overloaded) |
+| 504 | Gateway Timeout | When the server (acting as a gateway) could not get a response in a specified timeframe |
+
+#### JSONs
+
+JavaScript Object Notation (JSON) is a way of representing data in a very clean and easy to write way. It's commonly 
+compared to XML. You can see an in-depth description of everything on the 
+[JSON website](https://www.json.org/json-en.html).
+
+The JSON data format consists of three main data types; the `value`, `array`, and `object` types.
+
+##### Values
+
+Values are very simple, and there are only about seven main types: strings, numbers, objects, arrays, `true`, `false`,
+and `null`. 
+
+##### Arrays
+
+Arrays are a collection of `value`s commonly also referred to a vector, list, or sequence. See examples below:
+
+```json
+[]
+
+["a", "b", "c", "d", "e"]
+
+[[], ["a"], ["b", "c"]]
+
+[1, 2, ["a"], "b", true, false, null, [[[[]]]]]
+```
+
+##### Objects
+
+Objects are similar to arrays, however, they have keys which are assigned to each value. They're also commonly known as  
+serialized versions of an object, record, struct, dictionary, hash (the data type kind), map, hash table, keyed list, or
+associative array if you've heard of any of those before.
```

Comment:
as a freshman, I don't think I knew what any of those terms mean
can be added to the dictionary!

Reply:
We'll throw those in tomorrow then!